File: core/src/test/java/meshu/core/l1/L1Test.java
package meshu.core.l1;

import org.junit.jupiter.api.Test;

import java.util.HexFormat;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * TESTVECTORS.md vector 6 (§3.6) and §9.1: L1 header, ACK bitmap, segmentation.
 */
class L1Test {

    private static final HexFormat HEX = HexFormat.of();

    // ------------------------------------------------------------ header

    @Test
    void singleFrameHeaderIs4Bytes() {
        FrameHeader h = FrameHeader.dataSingle(0x1234, 0);
        byte[] enc = h.encode();
        assertEquals(4, enc.length);
        assertEquals("10001234", HEX.formatHex(enc));
        assertFalse(h.multi());
    }

    @Test
    void fragmentedHeaderIs6Bytes() {
        FrameHeader h = FrameHeader.dataMulti(0x1234, 3, 8, 0, true);
        byte[] enc = h.encode();
        assertEquals(6, enc.length);
        // kind DATA (0x10) | version 0 → 0x10; flags MULTI|REQ_ACK = 0x03
        assertEquals(0x10, enc[0] & 0xFF);
        assertEquals(0x03, enc[1] & 0xFF);
        assertEquals(0x1234, h.msgId());
        assertTrue(h.multi());
        assertTrue(h.reqAck());
        assertEquals(3, h.seq());
        assertEquals(8, h.lastSeq());
    }

    @Test
    void attemptCounterOccupiesBits2And3() {
        FrameHeader h = FrameHeader.dataSingle(0x1234, 1);
        assertEquals(0x04, h.encode()[1] & 0xFF); // ATTEMPT=1 → flags 0x04
        assertEquals(1, h.attempt());
        assertEquals(3, FrameHeader.dataSingle(1, 3).attempt());
    }

    @Test
    void versionIsTopTwoBits() {
        // v1 = 0b00, so byte 0 for DATA is 0x10 (kind only).
        FrameHeader.Parsed p = FrameHeader.parse(new byte[]{0x10, 0x00, 0x12, 0x34}, 0);
        assertEquals(0, p.header().version());
        assertEquals(FrameKind.DATA, p.header().kind());
    }

    // ------------------------------------------------------------ vector 6: ACK bitmap

    /**
     * A 9-frame message; frames 3 and 7 lost. ACKBM = {@code 12 00 1234 09 7701}.
     */
    @Test
    void vector6AckBitmap() {
        byte[] bitmap = new byte[]{(byte) 0x77, 0x01};
        byte[] frame = Reassembler.ackbmFrame(0x1234, 9, bitmap);
        assertEquals("1200123409" + "7701", HEX.formatHex(frame));
        assertEquals(7, frame.length);
    }

    @Test
    void vector6MissingFrames() {
        byte[] bitmap = new byte[]{(byte) 0x77, 0x01};
        List<Integer> missing = FrameHeader.missingFromBitmap(9, bitmap);
        assertEquals(List.of(3, 7), missing);
    }

    @Test
    void ackBitmap255FrameMax() {
        // A 255-frame message needs a 32-byte bitmap (§3.6).
        int total = 255;
        byte[] bitmap = new byte[(total + 7) / 8];
        assertEquals(32, bitmap.length);
        // Mark all received → no missing.
        java.util.Arrays.fill(bitmap, (byte) 0xFF);
        bitmap[31] = (byte) 0x7F; // 255 frames → bit 254 max, byte 31 = 0b01111111
        assertTrue(FrameHeader.missingFromBitmap(total, bitmap).isEmpty());
    }

    // ------------------------------------------------------------ segmentation boundary (§3.2)

    @Test
    void singleFrameAtOrBelow161Bytes() {
        // body ≤ 161 → single frame.
        List<Segmenter.Frame> frames = Segmenter.segment(1, new byte[161], Segmenter.MTU_FLOOR);
        assertEquals(1, frames.size());
        assertFalse(frames.getFirst().header().multi());
        assertEquals(161 + 4, frames.getFirst().encode().length); // 4-byte header
    }

    @Test
    void fragmentedAbove161Bytes() {
        // body 162 → 2 fragments.
        List<Segmenter.Frame> frames = Segmenter.segment(1, new byte[162], Segmenter.MTU_FLOOR);
        assertEquals(2, frames.size());
        assertTrue(frames.get(0).header().multi());
        assertEquals(159, frames.get(0).body().length);
        assertEquals(3, frames.get(1).body().length);
        // Final frame requests ACK.
        assertTrue(frames.get(1).header().reqAck());
    }

    // ------------------------------------------------------------ §9.1 frame arithmetic

    @Test
    void frameArithmeticMatchesVector9() {
        // (l3Len, frames, wireBytes) from TESTVECTORS.md §9.
        int[][] cases = {
                {41, 1, 62},     // HELLO
                {42, 1, 63},     // MINT_QUOTE req
                {228, 2, 257},   // MINT_QUOTE resp
                {122, 1, 143},   // MINT req 1 sat
                {292, 2, 321},   // MINT req 1000 sat
                {361, 3, 396},   // MINT req 255 sat
                {905, 6, 958},   // MINT req 16777215 sat
                {203, 2, 232},   // MINT resp 6 sigs
                {200, 2, 229},   // MELT_QUOTE req
                {28, 1, 49},     // MELT_QUOTE resp
                {495, 4, 536},   // MELT req 3 inputs
                {826, 6, 879},   // MELT req 8 inputs
                {1354, 9, 1425}, // MELT req 16 inputs
                {478, 4, 519},   // SWAP req 3/8
                {809, 6, 862},   // SWAP req 8/8
                {335, 3, 370},   // CHECKSTATE req 10
                {7, 1, 28},      // CHECKSTATE resp 10
                {1655, 11, 1738},// CHECKSTATE req 50
                {17, 1, 38},     // CHECKSTATE resp 50
                {49, 1, 70},     // KEYSETS resp 3
                {2127, 14, 2228},// KEYS resp 64
                {3409, 22, 3558},// RESTORE req 100
                {18, 1, 39},     // RESTORE resp 0 hits
                {184, 2, 213},   // RESTORE resp 5 hits
                {680, 5, 727},   // RESTORE resp 20 hits
                {3320, 21, 3463},// RESTORE resp 100 hits
        };
        for (int[] c : cases) {
            int[] result = Segmenter.framesAndWire(c[0]);
            assertEquals(c[1], result[0], "frames for l3=" + c[0]);
            assertEquals(c[2], result[1], "wire for l3=" + c[0]);
        }
    }

    // ------------------------------------------------------------ reassembly (§3.5)

    @Test
    void reassemblesFragmentedMessage() {
        Reassembler r = new Reassembler();
        byte[] body = new byte[400]; // 3 fragments at 159 B
        for (int i = 0; i < body.length; i++) {
            body[i] = (byte) i;
        }
        List<Segmenter.Frame> frames = Segmenter.segment(7, body, Segmenter.MTU_FLOOR);
        assertEquals(3, frames.size());

        Reassembler.Result r0 = r.feed(frames.get(0).header(), frames.get(0).body());
        assertInstanceOf(Reassembler.Result.Pending.class, r0);
        Reassembler.Result r1 = r.feed(frames.get(1).header(), frames.get(1).body());
        assertInstanceOf(Reassembler.Result.Pending.class, r1);
        Reassembler.Result r2 = r.feed(frames.get(2).header(), frames.get(2).body());

        assertInstanceOf(Reassembler.Result.Complete.class, r2);
        assertArrayEquals(body, ((Reassembler.Result.Complete) r2).body());
    }

    @Test
    void singleFrameCompletesImmediately() {
        Reassembler r = new Reassembler();
        byte[] body = {1, 2, 3};
        Reassembler.Result res = r.feed(FrameHeader.dataSingle(1, 0), body);
        assertInstanceOf(Reassembler.Result.Complete.class, res);
        assertArrayEquals(body, ((Reassembler.Result.Complete) res).body());
    }

    @Test
    void duplicateSeqDiscardedIdempotently() {
        Reassembler r = new Reassembler();
        List<Segmenter.Frame> frames = Segmenter.segment(1, new byte[200], Segmenter.MTU_FLOOR);
        r.feed(frames.get(0).header(), frames.get(0).body());
        Reassembler.Result dup = r.feed(frames.get(0).header(), frames.get(0).body());
        assertInstanceOf(Reassembler.Result.Duplicate.class, dup);
    }

    @Test
    void inconsistentLastSeqAborts() {
        Reassembler r = new Reassembler();
        r.feed(FrameHeader.dataMulti(1, 0, 4, 0, false), new byte[10]);
        Reassembler.Result res = r.feed(FrameHeader.dataMulti(1, 1, 7, 0, false), new byte[10]);
        assertInstanceOf(Reassembler.Result.Inconsistent.class, res);
    }

    /** Review C3: seq beyond last_seq must abort cleanly, never throw. */
    @Test
    void seqBeyondLastSeqIsInconsistentNotCrash() {
        Reassembler r = new Reassembler();
        // First sight of msg_id with an out-of-range seq.
        Reassembler.Result fresh = r.feed(FrameHeader.dataMulti(2, 200, 4, 0, false), new byte[10]);
        assertInstanceOf(Reassembler.Result.Inconsistent.class, fresh);
        // And against existing consistent state.
        r.feed(FrameHeader.dataMulti(3, 0, 4, 0, false), new byte[10]);
        Reassembler.Result mid = r.feed(FrameHeader.dataMulti(3, 9, 4, 0, false), new byte[10]);
        assertInstanceOf(Reassembler.Result.Inconsistent.class, mid);
        // State for msg_id 3 was aborted; re-feeding valid frames starts clean.
        Reassembler.Result ok = r.feed(FrameHeader.dataMulti(3, 0, 4, 0, false), new byte[10]);
        assertInstanceOf(Reassembler.Result.Pending.class, ok);
    }

    @Test
    void reqAckProducesAckbm() {
        Reassembler r = new Reassembler();
        List<Segmenter.Frame> frames = Segmenter.segment(0x1234, new byte[162], Segmenter.MTU_FLOOR);
        // First frame (seq 0, no REQ_ACK) → pending, no ack.
        Reassembler.Result.Pending p0 = (Reassembler.Result.Pending) r.feed(frames.get(0).header(), frames.get(0).body());
        // Feed the final frame (seq 1, REQ_ACK) — but message completes, so check
        // a partial instead: feed seq 0 only, which has no REQ_ACK here.
        assertNull(p0.ack());
    }
}

File: wallet/src/l1/l1.test.ts
import { describe, expect, test } from 'vitest';
import {
  dataMulti,
  dataSingle,
  encodeFrame,
  encodeHeader,
  HEADER_MULTI,
  HEADER_SINGLE,
  missingFromBitmap,
  parseHeader,
} from './frame';
import { KIND_ACKBM, KIND_DATA } from './frameKind';
import { Reassembler, ackbmFrame } from './reassembler';
import { MTU_FLOOR, framesAndWire, segment, WINDOW } from './segmenter';

const hex = (b: Uint8Array) => Array.from(b).map(x => x.toString(16).padStart(2, '0')).join('');

describe('L1 header (§3.1)', () => {
  test('single-frame header is 4 bytes', () => {
    const h = dataSingle(0x1234, 0);
    expect(encodeHeader(h).length).toBe(4);
    expect(hex(encodeHeader(h))).toBe('10001234');
  });

  test('fragmented header is 6 bytes with MULTI|REQ_ACK', () => {
    const h = dataMulti(0x1234, 3, 8, 0, true);
    const enc = encodeHeader(h);
    expect(enc.length).toBe(6);
    expect(enc[0]).toBe(0x10);
    expect(enc[1]).toBe(0x03);
  });

  test('ATTEMPT occupies bits 2–3', () => {
    expect(encodeHeader(dataSingle(1, 1))[1]).toBe(0x04);
    expect(parseHeader(encodeFrame(dataSingle(1, 3), new Uint8Array(0)), 0).header.flags & 0x0c).toBe(0x0c);
  });

  test('version is top two bits (v1 → byte0 = kind only)', () => {
    const p = parseHeader(new Uint8Array([0x10, 0x00, 0x12, 0x34]), 0);
    expect(p.header.version).toBe(0);
    expect(p.header.kind).toBe(KIND_DATA);
  });
});

describe('vector 6: ACK bitmap (§3.6)', () => {
  test('9-frame message, frames 3 and 7 lost → 12 00 1234 09 7701', () => {
    const frame = ackbmFrame(0x1234, 9, new Uint8Array([0x77, 0x01]));
    expect(hex(frame)).toBe('1200123409' + '7701');
    expect(frame.length).toBe(7);
  });

  test('missing frames from bitmap', () => {
    expect(missingFromBitmap(9, new Uint8Array([0x77, 0x01]))).toEqual([3, 7]);
  });

  test('255-frame max bitmap', () => {
    const bitmap = new Uint8Array(32);
    bitmap.fill(0xff);
    bitmap[31] = 0x7f;
    expect(missingFromBitmap(255, bitmap)).toEqual([]);
  });
});

describe('segmentation boundary (§3.2)', () => {
  test('≤161 bytes → single frame with REQ_ACK', () => {
    const frames = segment(1, new Uint8Array(161), MTU_FLOOR);
    expect(frames.length).toBe(1);
    expect(frames[0]!.body.length).toBe(161);
    expect((frames[0]!.header.flags & 0x02) !== 0).toBe(true);
  });

  test('162 bytes → 2 fragments, final REQ_ACK', () => {
    const frames = segment(1, new Uint8Array(162), MTU_FLOOR);
    expect(frames.length).toBe(2);
    expect(frames[0]!.body.length).toBe(159);
    expect(frames[1]!.body.length).toBe(3);
    expect((frames[1]!.header.flags & 0x02) !== 0).toBe(true);
  });
});

describe('§9.1 frame arithmetic — TESTVECTORS table (26 ops)', () => {
  const cases: Array<[l3: number, frames: number, wire: number]> = [
    [41, 1, 62], [42, 1, 63], [228, 2, 257], [122, 1, 143], [292, 2, 321],
    [361, 3, 396], [905, 6, 958], [203, 2, 232], [200, 2, 229], [28, 1, 49],
    [495, 4, 536], [826, 6, 879], [1354, 9, 1425], [478, 4, 519], [809, 6, 862],
    [335, 3, 370], [7, 1, 28], [1655, 11, 1738], [17, 1, 38], [49, 1, 70],
    [2127, 14, 2228], [3409, 22, 3558], [18, 1, 39], [184, 2, 213],
    [680, 5, 727], [3320, 21, 3463],
  ];
  test.each(cases)('l3=%i → %i frames / %i wire bytes', (l3, frames, wire) => {
    const [f, w] = framesAndWire(l3);
    expect(f).toBe(frames);
    expect(w).toBe(wire);
  });
});

describe('reassembly (§3.5)', () => {
  const bodyOf = (n: number) => new Uint8Array(n).map((_, i) => i & 0xff);

  test('fragmented message reassembles', () => {
    const r = new Reassembler();
    const body = bodyOf(400);
    const frames = segment(7, body, MTU_FLOOR);
    expect(frames.length).toBe(3);
    const r0 = r.feed(frames[0]!.header, frames[0]!.body);
    expect(r0.kind).toBe('pending');
    const r1 = r.feed(frames[1]!.header, frames[1]!.body);
    expect(r1.kind).toBe('pending');
    const r2 = r.feed(frames[2]!.header, frames[2]!.body);
    expect(r2.kind).toBe('complete');
    expect(r2.kind === 'complete' && r2.body.length).toBe(body.length);
  });

  test('single frame completes immediately with ACK when requested', () => {
    const r = new Reassembler();
    const res = r.feed(dataSingle(1, 0, true), new Uint8Array([1, 2, 3]));
    expect(res.kind).toBe('complete');
    expect(res.kind === 'complete' && res.ack).not.toBeNull();
  });

  test('duplicate seq discarded idempotently', () => {
    const r = new Reassembler();
    const frames = segment(1, new Uint8Array(200), MTU_FLOOR);
    r.feed(frames[0]!.header, frames[0]!.body);
    expect(r.feed(frames[0]!.header, frames[0]!.body).kind).toBe('duplicate');
  });

  test('inconsistent last_seq aborts', () => {
    const r = new Reassembler();
    r.feed(dataMulti(1, 0, 4, 0, false), new Uint8Array(10));
    expect(r.feed(dataMulti(1, 1, 7, 0, false), new Uint8Array(10)).kind).toBe('inconsistent');
  });

  test('seq beyond last_seq aborts cleanly (review C3)', () => {
    const r = new Reassembler();
    expect(r.feed(dataMulti(2, 200, 4, 0, false), new Uint8Array(10)).kind).toBe('inconsistent');
    r.feed(dataMulti(3, 0, 4, 0, false), new Uint8Array(10));
    expect(r.feed(dataMulti(3, 9, 4, 0, false), new Uint8Array(10)).kind).toBe('inconsistent');
    expect(r.feed(dataMulti(3, 0, 4, 0, false), new Uint8Array(10)).kind).toBe('pending');
  });

  test('late duplicate after completion answers full ACKBM', () => {
    const r = new Reassembler();
    const frames = segment(1, new Uint8Array(162), MTU_FLOOR);
    r.feed(frames[0]!.header, frames[0]!.body);
    const final = r.feed(frames[1]!.header, frames[1]!.body);
    expect(final.kind).toBe('complete');
    // Simulate lost completing ACK: the sender retransmits seq 0.
    const dup = r.feed(frames[0]!.header, frames[0]!.body);
    expect(dup.kind).toBe('pending');
    expect(dup.kind === 'pending' && dup.ack).not.toBeNull();
  });

  test('completed cache is bounded at COMPLETED_CACHE_SIZE (review fix)', () => {
    const r = new Reassembler();
    const frames = segment(1, new Uint8Array(162), MTU_FLOOR);
    r.feed(frames[0]!.header, frames[0]!.body);
    r.feed(frames[1]!.header, frames[1]!.body);
    // In cache: a late duplicate gets the full-ACKBM answer…
    const cached = r.feed(frames[0]!.header, frames[0]!.body);
    expect(cached.kind === 'pending' && cached.ack).not.toBeNull();
    // …until 64 newer completions evict it; then it is a fresh reassembly.
    for (let id = 2; id <= 66; id++) r.feed(dataSingle(id, 0), new Uint8Array(1));
    const dup = r.feed(frames[0]!.header, frames[0]!.body);
    expect(dup.kind).toBe('pending');
    expect(dup.kind === 'pending' && dup.ack).toBeNull();
  });
});
